animation_nodes/nodes/mesh/points_scatter.pyx

```diff
@@ -0,0 +1,165 @@
+    VirtualDoubleList,
+    PolygonIndicesList
+    )
+
+def randomPointsScatter(Vector3DList vertices, PolygonIndicesList polygons, VirtualDoubleList weights,
+                        bint useWeightForDensity, Py_ssize_t seed, Py_ssize_t pointAmount):
+    cdef FloatList triWeights, triAreas
+    cdef Py_ssize_t polyAmount
+    polyAmount, triAreas, triWeights = calculateTriangleWeightsAreas(vertices, polygons, weights)
+
+    cdef LongList distribution
+    cdef Py_ssize_t expectedAmount
+    distribution, expectedAmount = distributionOfPoints(polyAmount, triAreas, triWeights, useWeightForDensity,
+                                                        seed, pointAmount)
+
+    return sampleRandomPoints(vertices, polygons, distribution, seed, polyAmount, expectedAmount)
+
+@cython.cdivision(True)
+cdef calculateTriangleWeightsAreas(Vector3DList vertices, PolygonIndicesList polygons, VirtualDoubleList weights):
+    cdef Py_ssize_t vertexAmount = vertices.length
+    cdef float weightMax = 1
+    cdef float weight
+    cdef Py_ssize_t i
+
+    for i in range(vertexAmount):
+        weight = weights.get(i)
+        if weight > weightMax: weightMax = weight
+
+    cdef UIntegerList polyIndices = polygons.indices
+    cdef UIntegerList polyStarts = polygons.polyStarts
+    cdef Py_ssize_t polyAmount = polygons.getLength()
+    cdef FloatList triAreas = FloatList(length = polyAmount)
+    cdef FloatList triWeights = FloatList(length = polyAmount)
+    cdef Py_ssize_t start, polyIndex1, polyIndex2, polyIndex3
+
+    for i in range (polyAmount):
+        start = polyStarts.data[i]
+        polyIndex1 = polyIndices.data[start]
+        polyIndex2 = polyIndices.data[start + 1]
+        polyIndex3 = polyIndices.data[start + 2]
+
+        triAreas.data[i] = triangleArea(vertices.data[polyIndex1], vertices.data[polyIndex2], vertices.data[polyIndex3])
+        triWeights.data[i] = (weights.get(polyIndex1) + weights.get(polyIndex2) + weights.get(polyIndex3)) / weightMax / 3.0
+
+    return polyAmount, triAreas, triWeights
+
+@cython.cdivision(True)
+cdef float triangleArea(Vector3 v1, Vector3 v2, Vector3 v3):
+    cdef Vector3 vs1, vs2, vc
+    vs1.x = v1.x - v3.x
+    vs1.y = v1.y - v3.y
+    vs1.z = v1.z - v3.z
+
+    vs2.x = v2.x - v3.x
+    vs2.y = v2.y - v3.y
+    vs2.z = v2.z - v3.z
+
+    vc.x = vs1.y * vs2.z - vs1.z * vs2.y
+    vc.y = vs1.z * vs2.x - vs1.x * vs2.z
+    vc.z = vs1.x * vs2.y - vs1.y * vs2.x
+    return sqrt(vc.x * vc.x + vc.y * vc.y + vc.z * vc.z) / 2.0
+
+@cython.cdivision(True)
+cdef distributionOfPoints(Py_ssize_t polyAmount, FloatList triAreas, FloatList triWeights,
+                          bint useWeightForDensity, Py_ssize_t seed, Py_ssize_t pointAmount):
+    # Initial distribution of points.
+    cdef double triAreaTotal = 0
+    cdef Py_ssize_t i
+    for i in range(polyAmount):
+        triAreaTotal += triAreas.data[i]
+
+    if triAreaTotal == 0: triAreaTotal = 1
+    triAreaTotal = pointAmount / triAreaTotal
```

Comment:
Don't reuse variables here. Just create a new variable and name it something appropriate. `pointsPerUnitArea` maybe?

Reply:
Okay.